File: lc/A033_Search_in_Rotated_Sorted_Array.js
/**
Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

(i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).

You are given a target value to search. If found in the array return its index, otherwise return -1.

You may assume no duplicate exists in the array.
 */

var search = function(nums, target) {
    let low = 0, high = nums.length - 1;

    while (low < high) {
        let mid = Math.floor((low + high) / 2);
        
        if (nums[mid] === target) return mid;
        
        if (nums[mid] > nums[high]) {
            if (nums[mid] > target && nums[low] <= target) high = mid;
            else low = mid + 1;
        } else if (num[mid] < nums[high]) {
            if (nums[mid] < target && nums[high] >= target) low = mid + 1;
            else high = mid;
        } else {
            high--;
        }
    }

    return nums[low] === target ? low : - 1;
};

const s2 = (nums, target) => {
    let lo = 0, hi = nums.length - 1;
    while (lo <= hi) {
        let mid = Math.floor((lo + hi)/2);
        if (nums[mid] === target) return mid;
        if (nums[mid] >= nums[lo]) {
            if (target < nums[mid] && target >= nums[lo]) {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }

        if (nums[mid] <= nums[hi]) {
            if (target > nums[mid] && target <= nums[hi]) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
    }
    return -1;
}

console.log(search([1], 2));

File: java/src/leetcode/problems/A033_Search_in_Rotated_Sorted_Array.java
package leetcode.problems;

public class A033_Search_in_Rotated_Sorted_Array {

    public int search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        int lo = 0;
        int hi = nums.length - 1;

        while (lo + 1 < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                if (nums[mid] <= nums[lo] || target >= nums[lo]) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            } else {
                if (nums[mid] > nums[lo] || target < nums[lo]) {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }
        }

        if (nums[lo] == target) {
            return lo;
        }

        if (lo < nums.length - 1 && nums[lo + 1] == target) {
            return lo + 1;
        }

        return -1;
    }
}
